modules/dreamview/backend/data_collection_monitor/data_collection_monitor.cc

```diff
@@ -124,23 +125,66 @@ void DataCollectionMonitor::LoadConfiguration() {
       << "Unable to parse data collection configuration from file "
       << data_collection_config_path;
 
-  for (const auto& scenario_iter : data_collection_table_.scenario()) {
+  ConstructCategories();
+  for (const auto& scenario_iter : scenario_to_categories_) {
     const std::string& scenario_name = scenario_iter.first;
-    const Scenario& scenario = scenario_iter.second;
+    const auto& categories = scenario_iter.second;
 
-    for (const auto& category_iter : scenario.category()) {
+    for (const auto& category_iter : categories) {
       const std::string& category_name = category_iter.first;
-      const Category& category = category_iter.second;
-
       category_consecutive_frame_count_[scenario_name][category_name] = 0.0;
       category_frame_count_[scenario_name][category_name] = 0.0;
-      current_progress_json_[scenario_name][category.description()] = 0.0;
+      current_progress_json_[scenario_name][category_name] = 0.0;
     }
   }
 
   ADEBUG << "Configuration loaded.";
 }
 
+void DataCollectionMonitor::ConstructCategories() {
+  scenario_to_categories_.clear();
+
+  for (const auto& scenario_iter : data_collection_table_.scenario()) {
+    const std::string& scenario_name = scenario_iter.first;
+    const Scenario& scenario = scenario_iter.second;
+
+    Category category;
+    ConstructCategoriesHelper(scenario, 0, "", category,
+                              &scenario_to_categories_[scenario_name]);
```

Comment:
maybe pass in the scenario_name and let the helper access scenario_to_categories_ directly?